code/modules/mob/dead/observer/observer.dm

```diff
@@ -460,6 +456,106 @@ This is the proc mobs get to turn into a ghost. Forked from ghostize due to comp
 		host.ckey = src.ckey
 		host << "<span class='info'>You are now a mouse. Try to avoid interaction with players, and do not give hints away that you are more than a simple rodent.</span>"
 
+/proc/find_mouse_spawnpoint(var/ZLevel)
+	//This function will attempt to find a good spawnpoint for mice, and prevent them from spawning in closed vent systems with no escape
+	//It does this by bruteforce: Picks a random vent, tests if it has enough connections, if not, repeat
+	//Continues either until a valid one is found (in which case we return it), or until we hit a limit on attempts..
+	//If we hit the limit without finding a valid one, then the best one we found is selected
+
+	var/list/found_vents = list()
+	for(var/obj/machinery/atmospherics/unary/vent_pump/v in machines)
+		if(!v.welded && v.z == ZLevel)
+			found_vents.Add(v)
+
+	var/attempts = 0
+	var/max_attempts = 20
+	var/target_connections = 30//Any vent with at least this many connections is good enough
+
+	var/obj/machinery/atmospherics/unary/vent_pump/bestvent = null
+	var/best_connections = 0
+	while (attempts <= max_attempts)
+		attempts++
+		var/obj/machinery/atmospherics/unary/vent_pump/testvent = pick(found_vents)
+
+		if (testvent.welded)
+			continue
+
+		if (!testvent.network)//this prevents runtime errors
+			continue
+
+		var/turf/T = get_turf(testvent)
+
+
+
+		//We test the environment of the tile, to see if its habitable for a mouse
+		//-----------------------------------
+		var/atmos_suitable = 1
+
+		var/maxtemp = 390
+		var/mintemp = 210
+		var/min_oxy = 5
+		var/max_phoron = 1
+		var/max_co2 = 5
+		var/min_pressure = 80
+
+		var/datum/gas_mixture/Environment = T.return_air()
+		if(Environment)
+
+			if(Environment.temperature > maxtemp)
+				atmos_suitable = 0
+			else if (Environment.temperature < mintemp)
+				atmos_suitable = 0
+			else if(Environment.gas["oxygen"] < min_oxy)
+				atmos_suitable = 0
+			else if(Environment.gas["phoron"] > max_phoron)
+				atmos_suitable = 0
+			else if(Environment.gas["carbon_dioxide"] > max_co2)
+				atmos_suitable = 0
+			else if(Environment.return_pressure() < min_pressure)
+				atmos_suitable = 0
+		else
+			atmos_suitable = 0
+
+		if (!atmos_suitable)
+			continue
+		//----------------------
+
+
+
+
+		//Now we test the vent connections, and ensure the vent we spawn at is connected enough to give the mouse free movement
+		var/list/connections = list()
+		for(var/obj/machinery/atmospherics/unary/vent_pump/temp_vent in testvent.network.normal_members)
+			if(temp_vent.welded)
+				continue
+			if(temp_vent in testvent.loc)//Our testvent shouldn't count itself as a connection
+				continue
+			var/turf/T2 = get_turf(temp_vent)
+
+			var/i = 1
+			var/index = "[T2.loc.name]\[[i]\]"
+			while(index in connections)
+				i++
+				index = "[T2.loc.name]\[[i]\]"
+			connections[index] = temp_vent

```

Comment:
Is this complicated loop system necessary?

Since you're just looking to increase connections.len (and not access anything by index), you could just use
`connections += temp_vent` as every object is a unique object, it'll create an array like this:
`connections = list(1 = vent_object1, 2 = vent_object2)` where `connections[1] == vent_object1` and so on.

Though, I'm unsure why a list is even necessary. Could just use an integer and keep adding onto it per successful iteration.